File: application/src/main/java/com/example/server/mapper/BookingMapper.java
package com.example.server.mapper;

import com.example.api.dto.BookingCreateDto;
import com.example.api.dto.BookingDto;
import com.example.api.dto.ItemDto;
import com.example.api.dto.UserDto;
import com.example.api.dto.enums.BookingStatus;
import com.example.server.repository.entity.Booking;
import com.example.server.repository.entity.Item;
import com.example.server.repository.entity.User;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

@Component
@RequiredArgsConstructor
public class BookingMapper {

    private final ItemMapper itemMapper;
    private final UserMapper userMapper;

    public Booking toBooking(BookingCreateDto bookingCreateDto, Item item, User booker) {
        if ( bookingCreateDto == null && item == null && booker == null ) {
            return null;
        }

        Booking booking = new Booking();

        if ( bookingCreateDto != null ) {
            booking.setStart( bookingCreateDto.getStart() );
            booking.setEnd( bookingCreateDto.getEnd() );
        }
        booking.setItem( item );
        booking.setBooker( booker );
        booking.setStatus( BookingStatus.WAITING );

        return booking;
    }

    public BookingDto toBookingDto(Booking booking) {
        if ( booking == null ) {
            return null;
        }

        long id = 0L;
        LocalDateTime start = null;
        LocalDateTime end = null;
        ItemDto item = null;
        UserDto booker = null;
        BookingStatus status = null;

        if ( booking.getId() != null ) {
            id = booking.getId();
        }
        start = booking.getStart();
        end = booking.getEnd();
        item = itemMapper.toItemDto( booking.getItem() );
        booker = userMapper.toUserDto( booking.getBooker() );
        status = booking.getStatus();

        BookingDto bookingDto = new BookingDto( id, start, end, item, booker, status );

        return bookingDto;
    }

    public List<BookingDto> toBookingDto(List<Booking> bookings) {
        if ( bookings == null ) {
            return null;
        }

        List<BookingDto> list = new ArrayList<BookingDto>( bookings.size() );
        for ( Booking booking : bookings ) {
            list.add( toBookingDto( booking ) );
        }

        return list;
    }
}

File: application/src/main/kotlin/com/example/server/mapper/BookingMapper.kt
package com.example.server.mapper

import com.example.api.model.BookingCreateDto
import com.example.api.model.BookingDto
import com.example.api.model.BookingShortDto
import com.example.api.model.BookingStatus
import com.example.server.entity.Booking
import com.example.server.entity.Item
import com.example.server.entity.User
import java.time.LocalDateTime

fun BookingCreateDto.toEntity(item: Item, booker: User): Booking {
    return Booking(id = null, start = this.start, end = this.end, item = item, booker = booker, status = BookingStatus.WAITING)
}

fun Booking.toDto(): BookingDto {
    return BookingDto(id = this.id!!, start = this.start, end = this.end, item = this.item.toDto(), booker = this.booker.toDto(), status = this.status)
}

fun List<Booking>.toDto(): List<BookingDto> {
    return map { it.toDto() }
}

fun List<Booking>.findLast(): BookingShortDto? {
    val now = LocalDateTime.now()

    return this.stream()
        .sorted(Booking.BY_START_DESCENDING)
        .filter { booking: Booking -> booking.status == BookingStatus.APPROVED && booking.start.isBefore(now) }
        .findFirst()
        .map { booking: Booking -> BookingShortDto(booking.id!!, booking.start, booking.end, booking.booker.id!!) }
        .orElse(null)
}

fun List<Booking>.findNext(): BookingShortDto? {
    val now = LocalDateTime.now()

    return this.stream()
        .sorted(Booking.BY_START_ASCENDING)
        .filter { booking: Booking -> booking.status == BookingStatus.APPROVED && booking.start.isAfter(now) }
        .findFirst()
        .map { booking: Booking -> BookingShortDto(booking.id!!, booking.start, booking.end, booking.booker.id!!) }
        .orElse(null)
}
